File: src/android/CSSInjector.java
package com.vnkhoado.cordova.changeappinfo;

import android.graphics.Color;
import android.os.Handler;
import android.os.Looper;
import android.util.Base64;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;

public class CSSInjector extends CordovaPlugin {

    private static final String TAG = "CSSInjector";
    private static final String CSS_FILE_PATH = "www/assets/cdn-styles.css";
    private static final String CONFIG_FILE_PATH = "www/cordova-build-config.json";
    
    private String cachedCSS = null;
    private JSONObject cachedConfig = null;
    private Handler handler;
    private String backgroundColor = null;
    private boolean initialInjectionDone = false;

    @Override
    public void pluginInitialize() {
        super.pluginInitialize();
        
        // Read WEBVIEW_BACKGROUND_COLOR from preferences
        String bgColor = preferences.getString("WEBVIEW_BACKGROUND_COLOR", null);
        if (bgColor == null || bgColor.isEmpty()) {
            bgColor = preferences.getString("BackgroundColor", null);
        }
        if (bgColor == null || bgColor.isEmpty()) {
            bgColor = preferences.getString("SplashScreenBackgroundColor", null);
        }
        
        backgroundColor = bgColor;
        
        // Set WebView and Activity background
        final String finalBgColor = bgColor;
        cordova.getActivity().runOnUiThread(() -> {
            if (finalBgColor != null && !finalBgColor.isEmpty()) {
                try {
                    int color = parseHexColor(finalBgColor);
                    cordova.getActivity().getWindow().setBackgroundDrawable(
                        new android.graphics.drawable.ColorDrawable(color)
                    );
                    if (webView != null && webView.getView() != null) {
                        webView.getView().setBackgroundColor(color);
                    }
                    android.util.Log.d(TAG, "Background set to: " + finalBgColor);
                } catch (IllegalArgumentException e) {
                    android.util.Log.e(TAG, "Invalid color: " + finalBgColor, e);
                }
            }
        });
        
        // Pre-load CSS and config
        cachedCSS = readCSSFromAssets();
        cachedConfig = readConfigFromAssets();
        
        handler = new Handler(Looper.getMainLooper());
        
        android.util.Log.d(TAG, "CSSInjector initialized");
    }

    @Override
    public void onResume(boolean multitasking) {
        super.onResume(multitasking);
        
        if (!initialInjectionDone) {
            handler.postDelayed(() -> {
                injectAllContent();
                initialInjectionDone = true;
            }, 500);
        }
    }

    /**
     * Inject all content: config, background CSS, CDN CSS
     */
    private void injectAllContent() {
        // 1. Inject build config into window FIRST (most important)
        injectBuildConfig();
        
        // 2. Inject background color CSS
        if (backgroundColor != null && !backgroundColor.isEmpty()) {
            handler.postDelayed(() -> injectBackgroundColorCSS(backgroundColor), 100);
        }
        
        // 3. Inject CDN CSS
        handler.postDelayed(() -> injectCSSIntoWebView(), 200);
        
        android.util.Log.d(TAG, "All content injection scheduled");
    }

    /**
     * Inject build config from JSON file into window variable
     */
    private void injectBuildConfig() {
        cordova.getActivity().runOnUiThread(() -> {
            try {
                JSONObject config = cachedConfig;
                if (config == null) {
                    config = readConfigFromAssets();
                    cachedConfig = config;
                }
                
                if (config == null) {
                    android.util.Log.w(TAG, "No config found, skipping injection");
                    return;
                }
                
                // Add background color to config
                if (backgroundColor != null && !backgroundColor.isEmpty()) {
                    config.put("backgroundColor", backgroundColor);
                }
                
                // Convert to JSON string and escape for JavaScript
                String configJSON = config.toString();
                String escapedJSON = configJSON
                    .replace("\\", "\\\\")
                    .replace("'", "\\'")
                    .replace("\"", "\\\"")
                    .replace("\n", "\\n");
                
                CordovaWebView cordovaWebView = this.webView;
                if (cordovaWebView != null) {
                    String javascript = "(function() {" +
                        "  try {" +
                        "    var config = JSON.parse(\"" + escapedJSON + "\");" +
                        "    window.CORDOVA_BUILD_CONFIG = config;" +
                        "    window.AppConfig = config;" +
                        "    console.log('[Native] Build config injected:', config);" +
                        "    " +
                        "    if (typeof CustomEvent !== 'undefined') {" +
                        "      window.dispatchEvent(new CustomEvent('cordova-config-ready', { detail: config }));" +
                        "    }" +
                        "  } catch(e) {" +
                        "    console.error('[Native] Config injection failed:', e);" +
                        "  }" +
                        "})();";
                    
                    cordovaWebView.loadUrl("javascript:" + javascript);
                    android.util.Log.d(TAG, "Build config injected: " + config.toString());
                }
            } catch (Exception e) {
                android.util.Log.e(TAG, "Failed to inject build config", e);
            }
        });
    }

    /**
     * Read config JSON from assets
     */
    private JSONObject readConfigFromAssets() {
        StringBuilder content = new StringBuilder();
        
        try {
            InputStream inputStream = cordova.getActivity().getAssets().open(CONFIG_FILE_PATH);
            BufferedReader reader = new BufferedReader(
                new InputStreamReader(inputStream, StandardCharsets.UTF_8)
            );
            
            String line;
            while ((line = reader.readLine()) != null) {
                content.append(line);
            }
            
            reader.close();
            inputStream.close();
            
            return new JSONObject(content.toString());
            
        } catch (IOException e) {
            android.util.Log.w(TAG, "Config file not found: " + CONFIG_FILE_PATH);
            return null;
        } catch (JSONException e) {
            android.util.Log.e(TAG, "Failed to parse config JSON", e);
            return null;
        }
    }

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        if (action.equals("injectCSS")) {
            injectCSSIntoWebView();
            callbackContext.success("CSS injected");
            return true;
        } else if (action.equals("getConfig")) {
            // Allow JS to get config on demand
            JSONObject config = cachedConfig != null ? cachedConfig : readConfigFromAssets();
            if (config != null) {
                callbackContext.success(config);
            } else {
                callbackContext.error("Config not available");
            }
            return true;
        }
        return false;
    }

    private int parseHexColor(String hexColor) throws IllegalArgumentException {
        String hex = hexColor.trim();
        if (hex.startsWith("#")) {
            hex = hex.substring(1);
        }
        if (hex.length() != 6 && hex.length() != 8) {
            throw new IllegalArgumentException("Hex must be 6 or 8 chars");
        }
        try {
            if (hex.length() == 6) {
                return Color.parseColor("#FF" + hex);
            } else {
                return Color.parseColor("#" + hex);
            }
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException("Invalid color: " + hexColor, e);
        }
    }

    private void injectBackgroundColorCSS(final String bgColor) {
        cordova.getActivity().runOnUiThread(() -> {
            try {
                CordovaWebView cordovaWebView = this.webView;
                if (cordovaWebView != null) {
                    String css = "html, body { background-color: " + bgColor + " !important; margin: 0; padding: 0; }";
                    String javascript = "(function() {" +
                        "  function inject() {" +
                        "    try {" +
                        "      var target = document.head || document.getElementsByTagName('head')[0] || document.documentElement;" +
                        "      if (!target) {" +
                        "        setTimeout(inject, 100);" +
                        "        return;" +
                        "      }" +
                        "      var s = document.getElementById('cordova-bg');" +
                        "      if (s) s.remove();" +
                        "      s = document.createElement('style');" +
                        "      s.id = 'cordova-bg';" +
                        "      s.textContent = '" + css.replace("'", "\\'") + "';" +
                        "      target.appendChild(s);" +
                        "      console.log('[Native] Background CSS: " + bgColor + "');" +
                        "    } catch(e) { console.error('[Native] BG failed:', e); }" +
                        "  }" +
                        "  if (document.readyState === 'loading') {" +
                        "    document.addEventListener('DOMContentLoaded', inject);" +
                        "  } else {" +
                        "    inject();" +
                        "  }" +
                        "})();";
                    
                    cordovaWebView.loadUrl("javascript:" + javascript);
                }
            } catch (Exception e) {
                android.util.Log.e(TAG, "Background CSS failed", e);
            }
        });
    }

    private void injectCSSIntoWebView() {
        cordova.getActivity().runOnUiThread(() -> {
            try {
                String cssContent = cachedCSS;
                if (cssContent == null || cssContent.isEmpty()) {
                    cssContent = readCSSFromAssets();
                    cachedCSS = cssContent;
                }
                
                if (cssContent != null && !cssContent.isEmpty()) {
                    CordovaWebView cordovaWebView = this.webView;
                    if (cordovaWebView != null) {
                        String javascript = buildCSSInjectionScript(cssContent);
                        cordovaWebView.loadUrl("javascript:" + javascript);
                        android.util.Log.d(TAG, "CDN CSS injected (" + cssContent.length() + " bytes)");
                    }
                }
            } catch (Exception e) {
                android.util.Log.e(TAG, "CSS injection failed", e);
            }
        });
    }

    private String readCSSFromAssets() {
        StringBuilder cssContent = new StringBuilder();
        try {
            InputStream inputStream = cordova.getActivity().getAssets().open(CSS_FILE_PATH);
            BufferedReader reader = new BufferedReader(
                new InputStreamReader(inputStream, StandardCharsets.UTF_8)
            );
            String line;
            while ((line = reader.readLine()) != null) {
                cssContent.append(line).append("\n");
            }
            reader.close();
            inputStream.close();
        } catch (IOException e) {
            android.util.Log.e(TAG, "Failed to read CSS", e);
            return null;
        }
        return cssContent.toString();
    }

    private String buildCSSInjectionScript(String cssContent) {
        try {
            byte[] cssBytes = cssContent.getBytes(StandardCharsets.UTF_8);
            String base64CSS = Base64.encodeToString(cssBytes, Base64.NO_WRAP);
            
            return "(function() {" +
                   "  function inject() {" +
                   "    try {" +
                   "      var target = document.head || document.getElementsByTagName('head')[0] || document.documentElement;" +
                   "      if (!target) {" +
                   "        setTimeout(inject, 100);" +
                   "        return;" +
                   "      }" +
                   "      if (!document.getElementById('cdn-styles')) {" +
                   "        var b64 = '" + base64CSS + "';" +
                   "        var css = decodeURIComponent(escape(atob(b64)));" +
                   "        var s = document.createElement('style');" +
                   "        s.id = 'cdn-styles';" +
                   "        s.textContent = css;" +
                   "        target.appendChild(s);" +
                   "        console.log('[Native] CDN CSS loaded');" +
                   "      }" +
                   "    } catch(e) { console.error('[Native] CDN CSS failed:', e); }" +
                   "  }" +
                   "  if (document.readyState === 'loading') {" +
                   "    document.addEventListener('DOMContentLoaded', inject);" +
                   "  } else {" +
                   "    inject();" +
                   "  }" +
                   "})();";
        } catch (Exception e) {
            return buildFallbackInjectionScript(cssContent);
        }
    }

    private String buildFallbackInjectionScript(String cssContent) {
        String escaped = cssContent
            .replace("\\", "\\\\")
            .replace("'", "\\'")
            .replace("\"", "\\\"")
            .replace("\n", "\\n")
            .replace("\r", "")
            .replace("\t", "\\t");
        
        return "(function() {" +
               "  function inject() {" +
               "    try {" +
               "      var target = document.head || document.getElementsByTagName('head')[0] || document.documentElement;" +
               "      if (!target) {" +
               "        setTimeout(inject, 100);" +
               "        return;" +
               "      }" +
               "      if (!document.getElementById('cdn-styles')) {" +
               "        var s = document.createElement('style');" +
               "        s.id = 'cdn-styles';" +
               "        s.textContent = '" + escaped + "';" +
               "        target.appendChild(s);" +
               "        console.log('[Native] CDN CSS loaded');" +
               "      }" +
               "    } catch(e) { console.error('[Native] CSS failed:', e); }" +
               "  }" +
               "  if (document.readyState === 'loading') {" +
               "    document.addEventListener('DOMContentLoaded', inject);" +
               "  } else {" +
               "    inject();" +
               "  }" +
               "})();";
    }
}

File: src/ios/CSSInjector.swift
import Foundation
import WebKit
import UIKit

@objc(CSSInjector)
class CSSInjector: CDVPlugin {
    
    private static let CSS_FILE_PATH = "www/assets/cdn-styles.css"
    private static let CONFIG_FILE_PATH = "www/cordova-build-config.json"
    private var cachedCSS: String?
    private var cachedConfig: [String: Any]?
    
    override func pluginInitialize() {
        super.pluginInitialize()
        
        // Read WEBVIEW_BACKGROUND_COLOR from preferences (with fallbacks)
        var bgColor: String?
        if let color = self.commandDelegate.settings["webview_background_color"] as? String {
            bgColor = color
        } else if let color = self.commandDelegate.settings["backgroundcolor"] as? String {
            bgColor = color
        } else if let color = self.commandDelegate.settings["splashscreenbackgroundcolor"] as? String {
            bgColor = color
        }
        
        if let color = bgColor {
            setWebViewBackgroundColor(colorString: color)
            injectBackgroundColorCSS(colorString: color)
        }
        
        // Pre-load CSS and config
        cachedCSS = readCSSFromBundle()
        cachedConfig = readConfigFromBundle()
        
        // Inject config, CSS after a short delay to ensure WebView is ready
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
            self.injectBuildConfig()  // Inject config first
            self.injectCSSIntoWebView()
        }
        
        print("[CSSInjector] Plugin initialized")
    }
    
    @objc(injectCSS:)
    func injectCSS(command: CDVInvokedUrlCommand) {
        injectCSSIntoWebView()
        
        let pluginResult = CDVPluginResult(
            status: CDVCommandStatus_OK,
            messageAs: "CSS injected"
        )
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(getConfig:)
    func getConfig(command: CDVInvokedUrlCommand) {
        var config = cachedConfig
        if config == nil {
            config = readConfigFromBundle()
            cachedConfig = config
        }
        
        if let configDict = config {
            let pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: configDict
            )
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        } else {
            let pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Config not available"
            )
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        }
    }
    
    // MARK: - Build Config Injection
    
    /**
     * Read config JSON from bundle
     */
    private func readConfigFromBundle() -> [String: Any]? {
        guard let bundlePath = Bundle.main.path(forResource: "www", ofType: nil) else {
            print("[CSSInjector] www bundle path not found")
            return nil
        }
        
        let configPath = (bundlePath as NSString).appendingPathComponent("cordova-build-config.json")
        
        do {
            let jsonData = try Data(contentsOf: URL(fileURLWithPath: configPath))
            let config = try JSONSerialization.jsonObject(with: jsonData, options: []) as? [String: Any]
            return config
        } catch {
            print("[CSSInjector] Failed to read config: \(error.localizedDescription)")
            return nil
        }
    }
    
    /**
     * Inject build config into window
     */
    private func injectBuildConfig() {
        DispatchQueue.main.async {
            guard let wkWebView = self.webView as? WKWebView else {
                print("[CSSInjector] WKWebView not available for config injection")
                return
            }
            
            var config = self.cachedConfig
            if config == nil {
                config = self.readConfigFromBundle()
                self.cachedConfig = config
            }
            
            guard var configDict = config else {
                print("[CSSInjector] No config found, skipping injection")
                return
            }
            
            // Add background color if available
            if let bgColor = self.commandDelegate.settings["webview_background_color"] as? String {
                configDict["backgroundColor"] = bgColor
            } else if let bgColor = self.commandDelegate.settings["backgroundcolor"] as? String {
                configDict["backgroundColor"] = bgColor
            }
            
            // Convert to JSON string
            do {
                let jsonData = try JSONSerialization.data(withJSONObject: configDict, options: [])
                if let jsonString = String(data: jsonData, encoding: .utf8) {
                    let escapedJSON = jsonString
                        .replacingOccurrences(of: "\\", with: "\\\\")
                        .replacingOccurrences(of: "'", with: "\\'")
                        .replacingOccurrences(of: "\"", with: "\\\"")
                        .replacingOccurrences(of: "\n", with: "\\n")
                    
                    let javascript = """
                    (function() {
                        try {
                            var config = JSON.parse("\(escapedJSON)");
                            window.CORDOVA_BUILD_CONFIG = config;
                            window.AppConfig = config;
                            console.log('[Native iOS] Build config injected:', config);
                            
                            if (typeof CustomEvent !== 'undefined') {
                                window.dispatchEvent(new CustomEvent('cordova-config-ready', { detail: config }));
                            }
                        } catch(e) {
                            console.error('[Native iOS] Config injection failed:', e);
                        }
                    })();
                    """
                    
                    wkWebView.evaluateJavaScript(javascript) { (_, error) in
                        if let error = error {
                            print("[CSSInjector] Failed to inject config: \(error.localizedDescription)")
                        } else {
                            print("[CSSInjector] Build config injected successfully")
                        }
                    }
                }
            } catch {
                print("[CSSInjector] Failed to serialize config: \(error.localizedDescription)")
            }
        }
    }
    
    // MARK: - WebView Background Color
    
    /**
     * Set WebView background color to prevent white flash
     */
    private func setWebViewBackgroundColor(colorString: String) {
        DispatchQueue.main.async {
            guard let webView = self.webView as? WKWebView else {
                print("[CSSInjector] WebView not available for background color")
                return
            }
            
            // Parse hex color
            if let color = self.hexStringToUIColor(hex: colorString) {
                webView.backgroundColor = color
                webView.isOpaque = false
                webView.scrollView.backgroundColor = color
                print("[CSSInjector] WebView background set to: \(colorString)")
            } else {
                // Fallback to clear
                webView.backgroundColor = .clear
                webView.isOpaque = false
                print("[CSSInjector] Invalid color format, using clear: \(colorString)")
            }
        }
    }
    
    /**
     * Inject background color CSS into WebView at runtime
     */
    private func injectBackgroundColorCSS(colorString: String) {
        DispatchQueue.main.async {
            guard let wkWebView = self.webView as? WKWebView else {
                print("[CSSInjector] WKWebView not available for background CSS injection")
                return
            }
            
            let css = "html, body { background-color: \(colorString) !important; margin: 0; padding: 0; }"
            let escapedCSS = css.replacingOccurrences(of: "'", with: "\\'")
            let javascript = """
            (function() {
                try {
                    var existingStyle = document.getElementById('cordova-plugin-webview-bg');
                    if (existingStyle) { existingStyle.remove(); }
                    var style = document.createElement('style');
                    style.id = 'cordova-plugin-webview-bg';
                    style.textContent = '\(escapedCSS)';
                    (document.head || document.documentElement).appendChild(style);
                    console.log('Background CSS injected: \(colorString)');
                } catch(e) {
                    console.error('Background CSS injection failed:', e);
                }
            })();
            """
            
            wkWebView.evaluateJavaScript(javascript) { (_, error) in
                if let error = error {
                    print("[CSSInjector] Failed to inject background CSS: \(error.localizedDescription)")
                } else {
                    print("[CSSInjector] Background color CSS injected: \(colorString)")
                }
            }
        }
    }
    
    /**
     * Convert hex string to UIColor
     */
    private func hexStringToUIColor(hex: String) -> UIColor? {
        var hexSanitized = hex.trimmingCharacters(in: .whitespacesAndNewlines)
        hexSanitized = hexSanitized.replacingOccurrences(of: "#", with: "")
        
        var rgb: UInt64 = 0
        
        guard Scanner(string: hexSanitized).scanHexInt64(&rgb) else {
            return nil
        }
        
        let length = hexSanitized.count
        
        if length == 6 {
            let r = CGFloat((rgb & 0xFF0000) >> 16) / 255.0
            let g = CGFloat((rgb & 0x00FF00) >> 8) / 255.0
            let b = CGFloat(rgb & 0x0000FF) / 255.0
            return UIColor(red: r, green: g, blue: b, alpha: 1.0)
        } else if length == 8 {
            let r = CGFloat((rgb & 0xFF000000) >> 24) / 255.0
            let g = CGFloat((rgb & 0x00FF0000) >> 16) / 255.0
            let b = CGFloat((rgb & 0x0000FF00) >> 8) / 255.0
            let a = CGFloat(rgb & 0x000000FF) / 255.0
            return UIColor(red: r, green: g, blue: b, alpha: a)
        }
        
        return nil
    }
    
    // MARK: - CSS Injection
    
    /**
     * Read CSS from file and inject into WebView
     */
    private func injectCSSIntoWebView() {
        DispatchQueue.main.async {
            // Use cached CSS or read from bundle
            var cssContent = self.cachedCSS
            if cssContent == nil || cssContent!.isEmpty {
                cssContent = self.readCSSFromBundle()
                self.cachedCSS = cssContent
            }
            
            guard let css = cssContent, !css.isEmpty else {
                print("[CSSInjector] CSS file not found or empty")
                return
            }
            
            guard let wkWebView = self.webView as? WKWebView else {
                print("[CSSInjector] WKWebView not available")
                return
            }
            
            let javascript = self.buildCSSInjectionScript(cssContent: css)
            
            wkWebView.evaluateJavaScript(javascript) { (result, error) in
                if let error = error {
                    print("[CSSInjector] Failed to inject CSS: \(error.localizedDescription)")
                } else {
                    print("[CSSInjector] CSS injected successfully (\(css.count) bytes)")
                }
            }
        }
    }
    
    /**
     * Read CSS content from bundle www/assets/cdn-styles.css with UTF-8 encoding
     */
    private func readCSSFromBundle() -> String? {
        guard let bundlePath = Bundle.main.path(forResource: "www", ofType: nil) else {
            print("[CSSInjector] www bundle path not found")
            return nil
        }
        
        let cssPath = (bundlePath as NSString).appendingPathComponent("assets/cdn-styles.css")
        
        do {
            let cssContent = try String(contentsOfFile: cssPath, encoding: .utf8)
            return cssContent
        } catch {
            print("[CSSInjector] Failed to read CSS file: \(error.localizedDescription)")
            return nil
        }
    }
    
    /**
     * Build JavaScript code to inject CSS into page
     */
    private func buildCSSInjectionScript(cssContent: String) -> String {
        // Try Base64 encoding first (safest method)
        if let base64CSS = encodeToBase64(cssContent: cssContent) {
            return """
            (function() {
                try {
                    if (!document.getElementById('cdn-injected-styles')) {
                        var base64CSS = '\(base64CSS)';
                        var decodedCSS = decodeURIComponent(escape(atob(base64CSS)));
                        var style = document.createElement('style');
                        style.id = 'cdn-injected-styles';
                        style.textContent = decodedCSS;
                        (document.head || document.documentElement).appendChild(style);
                        console.log('CSS injected by native code (Base64)');
                    }
                } catch(e) {
                    console.error('CSS injection failed:', e);
                }
            })();
            """
        } else {
            // Fallback: Use manual escaping if Base64 fails
            return buildFallbackInjectionScript(cssContent: cssContent)
        }
    }
    
    /**
     * Encode CSS content to Base64
     */
    private func encodeToBase64(cssContent: String) -> String? {
        guard let data = cssContent.data(using: .utf8) else {
            print("[CSSInjector] Failed to encode CSS to UTF-8")
            return nil
        }
        
        return data.base64EncodedString(options: [])
    }
    
    /**
     * Fallback method using manual escaping
     */
    private func buildFallbackInjectionScript(cssContent: String) -> String {
        // Escape CSS content for use in JavaScript string
        let escapedCSS = cssContent
            .replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "'", with: "\\'")
            .replacingOccurrences(of: "\"", with: "\\\"")
            .replacingOccurrences(of: "\n", with: "\\n")
            .replacingOccurrences(of: "\r", with: "")
            .replacingOccurrences(of: "\t", with: "\\t")
        
        return """
        (function() {
            try {
                if (!document.getElementById('cdn-injected-styles')) {
                    var style = document.createElement('style');
                    style.id = 'cdn-injected-styles';
                    style.textContent = '\(escapedCSS)';
                    (document.head || document.documentElement).appendChild(style);
                    console.log('CSS injected by native code (escaped)');
                }
            } catch(e) {
                console.error('CSS injection failed:', e);
            }
        })();
        """
    }
}
